libraries/AP_Proximity/AP_Proximity_Boundary_3D.h

```diff
@@ -0,0 +1,141 @@
+
+    // add a distance to the boundary if it is shorter than any other provided distance since the last time the boundary was reset
+    // pitch and yaw are in degrees, distance is in meters
+    void add_distance(float pitch, float yaw, float distance);
+    void add_distance(float yaw, float distance) { add_distance(0, yaw, distance); }
+
+    // update boundary points used for simple avoidance based on a single sector and pitch distance changing
+    //   the boundary points lie on the line between sectors meaning two boundary points may be updated based on a single sector's distance changing
+    //   the boundary point is set to the shortest distance found in the two adjacent sectors, this is a conservative boundary around the vehicle
+    void update_boundary(Face face);
+
+    // update middle layer boundary points
+    void update_middle_boundary();
+
+    // reset boundary.  marks all distances as invalid
+    void reset();
+
+    // Reset this location, specified by Face object, back to default
+    // i.e Distance is marked as not-valid, and set to a large number.
+    void reset_face(Face face);
+
+    // get distance for a face.  returns true on success and fills in distance argument with distance in meters
+    bool get_distance(Face face, float &distance) const;
+
+    // Get the total number of obstacles 
+    uint8_t get_obstacle_count() const;
+
+    // Appropriate layer and sector are found from the passed obstacle_num
+    // This function then draws a line between this sector, and sector + 1 at the given layer
+    // Then returns the closest point on this line from vehicle, in body-frame. 
+    bool get_obstacle(uint8_t obstacle_num, Vector3f& vec_to_boundary) const;
+
+    // Appropriate layer and sector are found from the passed obstacle_num
+    // This function then draws a line between this sector, and sector + 1 at the given layer
+    // Then returns the closest point on this line from the segment that was passed, in body-frame.
+    // Used by GPS based Simple Avoidance  - for "brake mode" 
+    float distance_to_obstacle(uint8_t obstacle_num, const Vector3f& seg_start, const Vector3f& seg_end, Vector3f& closest_point) const;
+
+    // get distance and angle to closest object (used for pre-arm check)
+    //   returns true on success, false if no valid readings
+    bool get_closest_object(float& angle_deg, float &distance) const;
+
+    // get number of objects, angle and distance - used for non-GPS avoidance
+    uint8_t get_horizontal_object_count() const;
+    bool get_horizontal_object_angle_and_distance(uint8_t object_number, float& angle_deg, float &distance) const;
+
+    // sectors
+    const uint16_t _sector_middle_deg[PROXIMITY_NUM_SECTORS] {0, 45, 90, 135, 180, 225, 270, 315};    // middle angle of each sector
+    // layers
+    const int16_t _pitch_middle_deg[PROXIMITY_NUM_LAYERS] {-60, -30, 0, 30, 60};
```

Comment:
```suggestion
    const int16_t pitch_middle_deg(uint8_t layer) 
    {
        return ((PROXIMITY_NUM_LAYERS/2 * -PROXIMITY_PITCH_WIDTH_DEG) + layer * PROXIMITY_PITCH_WIDTH_DEG);
    }
```

.... or maybe just a static assert.  It's just you can't make it "easy" for people to change the number of layers with `PROXMITY_NUM_LAYERS` and leave this landmine down here for them to trip over.

You could avoid all the maths for every lookup by allocating-but-not-initialising _pitch_middle_deg, making it non-cost and filling it in in `init()`, of course.